core-api/src/main/java/com/optimizely/ab/OptimizelyUserContext.java

```diff
@@ -289,21 +297,33 @@ public void setQualifiedSegments(List<String> qualifiedSegments) {
         this.qualifiedSegments.addAll(qualifiedSegments);
     }
 
+    /**
+     * Fetch all qualified segments for the user context.
+     *
+     * The segments fetched will be saved and can be accessed at any time by calling {@link #getQualifiedSegments()}.
+     */
     public void fetchQualifiedSegments() {
-        fetchQualifiedSegments(false, false);
+        ODPManager odpManager = optimizely.getODPManager();
+        if (odpManager != null) {
+            synchronized (odpManager) {
+                setQualifiedSegments(odpManager.getSegmentManager().getQualifiedSegments(userId));
+            }
+        }
     }
 
-    public void fetchQualifiedSegments(@Nonnull Boolean ignoreCache, @Nonnull Boolean resetCache) {
+    /**
+     * Fetch all qualified segments for the user context.
+     *
+     * The segments fetched will be saved and can be accessed at any time by calling {@link #getQualifiedSegments()}.
+     *
+     * @param segmentOptions A set of options for fetching qualified segments.
+     */
+    public void fetchQualifiedSegments(@Nonnull List<ODPSegmentOption> segmentOptions) {
         ODPManager odpManager = optimizely.getODPManager();
         if (odpManager != null) {
-            List<ODPSegmentOption> segmentOptions = new ArrayList<>();
-            if (ignoreCache) {
-                segmentOptions.add(ODPSegmentOption.IGNORE_CACHE);
-            }
-            if (resetCache) {
-                segmentOptions.add(ODPSegmentOption.RESET_CACHE);
+            synchronized (odpManager) {
+                setQualifiedSegments(odpManager.getSegmentManager().getQualifiedSegments(userId, segmentOptions));
             }
```

Comment:
I will do this but i just wanted to avoid method drilling. I do not want us to end up creating multiple levels of simple wrappers that do not add any value in between. This is why i preferred exposing ODPEventManager and ODPSegmentManager directly as getters. One option is to add wrappers for each ODP related function in the `Optimizely` client. Another option is to let userContext use ODPManager directly so that we do not need to keep adding wrappers as we add functionality to ODPManager. This wrapper mechanism gets specially tricky when we provide multiple overrides of methods. I would suggest to keep it like this and let userContext access ODPManager through optimizely and call its methods direclty.. what do you say?

Reply:
@zashraf1985 it looks like a style preference with tradeoffs. I'll leave it to you.